readthedocs/projects/tasks.py

```diff
@@ -715,6 +723,123 @@ def save_build_config(self):
         })
         self.build['config'] = config
 
+    def store_build_artifacts(
+            self,
+            html=False,
+            localmedia=False,
+            search=False,
+            pdf=False,
+            epub=False,
+    ):
+        """
+        Save build artifacts to "storage" using Django's storage API
+
+        The storage could be local filesystem storage OR cloud blob storage
+        such as S3, Azure storage or Google Cloud Storage.
+
+        Remove build artifacts of types not included in this build (PDF, ePub, zip only).
+
+        This looks very similar to `move_files` and is intended to replace it!
+
+        :param html: whether to save HTML output
+        :param localmedia: whether to save localmedia (htmlzip) output
+        :param search: whether to save search artifacts
+        :param pdf: whether to save PDF output
+        :param epub: whether to save ePub output
+        """
+        if settings.RTD_BUILD_MEDIA_STORAGE:
+            log.info(
+                LOG_TEMPLATE.format(
+                    project=self.version.project.slug,
+                    version=self.version.slug,
+                    msg='Writing build artifacts to media storage',
+                ),
+            )
+
+            storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
+
+            types_to_copy = []
+            types_to_delete = []
+
+            # HTML media
+            if html:
+                types_to_copy.append(('html', self.config.doctype))
+
+            # Search media (JSON)
+            if search:
+                types_to_copy.append(('json', 'sphinx_search'))
+
+            if localmedia:
+                types_to_copy.append(('htmlzip', 'sphinx_localmedia'))
+            else:
+                types_to_delete.append('htmlzip')
+
+            if pdf:
+                types_to_copy.append(('pdf', 'sphinx_pdf'))
+            else:
+                types_to_delete.append('pdf')
+
+            if epub:
+                types_to_copy.append(('epub', 'sphinx_epub'))
+            else:
+                types_to_delete.append('epub')
+
+            for media_type, build_type in types_to_copy:
```

Comment:
nit:

This is a little confusing to me. What is `media_type` and `build_type`? I suppose the first one is the path name where we want to save it, and the later is the path name from the builder.

If so, where they come from? Isn't it possible to use a constant or get the name from the related class where this is defined?

Reply:
I tend to agree but I believe this is a larger refactor and should be handled in a separate PR. These constants **are not** defined somewhere convenient (like `readthedocs/doc_builder/constants.py`) and they are used in quite a few places outside of the scope of this PR like the doc builder backends themselves.